pkg/plugins/layout.go

```diff
@@ -0,0 +1,614 @@
+package plugins
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/mudler/yip/pkg/schema"
+	log "github.com/sirupsen/logrus"
+	"github.com/twpayne/go-vfs"
+)
+
+type Disk struct {
+	Device  string
+	SectorS uint
+	LastS   uint
+	Parts   []Partition
+}
+
+// We only manage sizes in sectors unit for the Partition struct and sgdisk wrapper
+type Partition struct {
+	Number     int
+	StartS     uint
+	SizeS      uint
+	PLabel     string
+	FileSystem string
+	FSLabel    string
+	Type       string
+}
+
+type GdiskCall struct {
+	dev       string
+	wipe      bool
+	parts     []*Partition
+	deletions []int
+	expand    bool
+	pretend   bool
+}
+
+type MkfsCall struct {
+	part       Partition
+	customOpts []string
+	dev        string
+}
+
+func Layout(s schema.Stage, fs vfs.FS, console Console) error {
+	if s.Layout.Device == nil {
+		return nil
+	}
+
+	dev, err := FindDiskFromPartitionLabel(s.Layout.Device.Label, console)
+	if err != nil {
+		log.Warningf("Exiting, disk not found:\n %s", err.Error())
+		return nil
+	}
+
+	changed := false
+
+	// Check there is a minimum of 32MiB of free space in disk
+	if !dev.CheckDiskFreeSpaceMiB(32, console) {
```

Comment:
small nit( feel free to ignore): maybe we should log a warn here, or in case it doesn't pass the user wouldn't know why it did nothing

Reply:
Yes, makes sense. Actually I am not super enthusiastic of this check, but it is the best method I could find to ensure the disk is already repartitioned (so we can boot and reboot and no changes on disk are applied or even attempted on reboot). Gonna add a warning here.